docs/csharp/proto2.md

```diff
@@ -0,0 +1,184 @@
+    optional Bar bar = 1;
+    required Baz baz = 2;
+}
+```
+```cs
+var foo = new Foo();
+Assert.IsNull(foo.Bar);
+Assert.False(foo.HasBar);
+foo.Bar = new Bar();
+Assert.True(foo.HasBar);
+foo.ClearBar();
+```
+
+### Messages with extension ranges
+
+Messages which define extension ranges implement the `IExtendableMessage` interface as shown below.
+See inline comments for more info.
+
+```cs
+public interface IExtendableMessage<T> : IMessage<T> where T : IExtendableMessage<T>
+{
+    // Gets the value of a single value extension. If the extension isn't present, this returns the default value.
+    TValue GetExtension<TValue>(Extension<T, TValue> extension);
+    // Gets the value of a repeated extension. If the extension hasn't been set, this returns null to prevent unnecessary allocations.
+    RepeatedField<TValue> GetExtension<TValue>(RepeatedExtension<T, TValue> extension);
+    // Gets the value of a repeated extension. This will initialize the value of the repeated field and will never return null.
+    RepeatedField<TValue> GetOrInitializeExtension<TValue>(RepeatedExtension<T, TValue> extension);
+    // Sets the value of the extension
+    void SetExtension<TValue>(Extension<T, TValue> extension, TValue value);
+    // Returns whether the extension is present in the message
+    bool HasExtension<TValue>(Extension<T, TValue> extension);
+    // Clears the value of the extension, removing it from the message
+    void ClearExtension<TValue>(Extension<T, TValue> extension);
+    // Clears the value of the repeated extension, removing it from the message. Calling GetExtension after this will always return null.
+    void ClearExtension<TValue>(RepeatedExtension<T, TValue> extension);
+}
+```
+
+### Extensions
+
+Extensions are generated in static containers like reflection classes and type classes. 
+For example for a file called `foo.proto` containing extensions in the file scope, a 
+`FooExtensions` class is created containing the extensions defined in the file scope.
+For easy access, this class can be used with `using static` to bring all extensions into scope.
+
+```proto
+option csharp_namespace = "FooBar";
+extend Foo {
+    optional Baz foo_ext = 124;
+}
+message Baz {
+    extend Foo {
+        repeated Baz repeated_foo_ext = 125;
+    }
+}
+```
+```cs
+public static partial class FooExtensions {
+    public static readonly Extension<Foo, Baz> FooExt = /* initialization */;
+}
+
+public partial class Baz {
+    public partial static class Extensions {
+        public static readonly RepeatedExtension<Foo, Baz> RepeatedFooExt = /* initialization */;
+    }
+}
+```
+```cs
+using static FooBar.FooExtensions;
+using static FooBar.Baz.Extensions;
+
+var foo = new Foo();
+foo.SetExtension(FooExt, new Baz());
+foo.GetOrInitializeExtension(RepeatedFooExt).Add(new Baz());
+```
+
+# APIs
+
+### Message initialization
+
+Checking message initialization is not handled automatically by the library, but can be done manually via the
```

Comment:
it might be useful to elaborate a bit more on what's meant by "message initialization". I think many users won't be familiar with what exactly that means.
we could also answer these questions: Also, what are the consequences of needing to do IsInitialized() manually? Is that the ways it behaves in other languages? 
If this is a limitation compared to other languages, it might be worth putting one sentence reasoning of that design decision.